### PR TITLE
sd-dhcp-client: gracefully ignore OFFER with Rapid Commit option

### LINKED ISSUE #29904: systemd-networkd: DHCPv4 won't acquire IP from Google Fiber on 255-rc1
### systemd version the issue has been seen with

255-rc1

### Used distribution

Kubuntu 24.04

### Linux kernel version used

6.6.0

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

systemd-networkd should acquire an IPv4 address as it did with previous versions.

### Unexpected behaviour you saw

systemd-networkd fails to acquire an IPv4 address.  The following log output is produced when debug logging is turned on:
```
systemd-networkd[228649]: wan: Requested configuring of the DHCPv4 client.
systemd-networkd[228649]: wan: DHCPv4 client: STARTED on ifindex 2
systemd-networkd[228649]: wan: DHCPv4 client is configured, acquiring DHCPv4 lease.
systemd-networkd[228649]: wan: DHCPv4 client: DISCOVER
systemd-networkd[228649]: wan: DHCPv4 client: State changed: initialization -> selecting
systemd-networkd[228649]: wan: DHCPv4 client: Received message from RAW socket, processing.
systemd-networkd[228649]: wan: DHCPv4 client: received OFFER with Rapid Commit option, ignoring
```
followed by an endless loop of:
```
systemd-networkd[228649]: wan: DHCPv4 client: DISCOVER
systemd-networkd[228649]: wan: DHCPv4 client: Received message from RAW socket, processing.
systemd-networkd[228649]: wan: DHCPv4 client: received OFFER with Rapid Commit option, ignoring
```
It would appear as if Google Fiber's DHCPv4 server is behaving in a standards-incompliant manner and sending an OFFER message with the Rapid Commit option, which breaks with the new Rapid Commit feature in 255.  This occurs regardless of whether RapidCommit is turned on or off.  I tried commenting out https://github.com/systemd/systemd/blob/main/src/libsystemd-network/sd-dhcp-client.c#L1612 through line 1614 and that resolves the issue.

The configuration of the relevant interface is:
```
[Match]
Name=wan

[Network]
DHCP=yes

[DHCPv4]
UseDNS=no
UseHostname=no
UseNTP=no
SendRelease=no

[DHCPv6]
WithoutRA=solicit
PrefixDelegationHint=::/56
UseDNS=no
UseNTP=no
SendRelease=no

[IPv6AcceptRA]
UseDNS=no
UseDomains=no

[Link]
RequiredForOnline=routable
RequiredFamilyForOnline=both
```

### Steps to reproduce the problem

Use systemd-networkd 255 to attempt to get a DHCPv4 lease from Google Fiber.

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #29904.

### CODE REVIEW

**Review by bluca — APPROVED** [review]

File: src/libsystemd-network/sd-dhcp-client.c [hunk 1]
```diff
@@ -1609,9 +1609,14 @@ static int client_parse_message(
...
                                                              "received rapid ACK without Rapid Commit option, ignoring.");
                 } else if (r == DHCP_OFFER) {
-                        if (lease->rapid_commit)
-                                return log_dhcp_client_errno(client, SYNTHETIC_ERRNO(ENOMSG),
-                                                             "received OFFER with Rapid Commit option, ignoring");
+                        if (lease->rapid_commit) {
+                                /* Some RFC incompliant servers provides an OFFER with a rapid commit option.
+                                 * See https://github.com/systemd/systemd/issues/29904.
+                                 * Let's support such servers gracefully. */
+                                log_dhcp_client_errno(client, SYNTHETIC_ERRNO(ENOMSG),
```
Comment (YHNdnzj) [inline src/libsystemd-network/sd-dhcp-client.c]: Drop the synthetic errno since we're not using the return value?

Comment (yuwata) [reply src/libsystemd-network/sd-dhcp-client.c]: Nice catch. Fixed.

**Review by keszybz — COMMENTED** [review]
LGTM too.

**Comment by mamarley** [comment]
I see this has already been merged now, but it does solve the problem.  Thanks!

**Comment by yuwata** [comment]
@mamarley Thank you for testing the patch!

**Pull request merged**